BootloaderCorePkg/Tools/GenContainer.py

```diff
@@ -328,76 +362,133 @@ def dump (self):
+			if is_last_entry:
+				raise Exception ("'%s' must be the last entry in layout for monolithic signing!" % mono_sig)
 			if compress_alg == '':
 				compress_alg = 'Dummy'
 			if auth_type == '':
 				auth_type = 'NONE'
+
+			# build a component entry
 			component = COMPONENT_ENTRY ()
 			component.name      = name.encode()
-			component.alignment = self.header.alignment.bit_length() - 1
+			if alignment == 0:
+				component.alignment = self.header.alignment.bit_length() - 1
+			else:
+				component.alignment = alignment.bit_length() - 1
+			component.attribute = 0
 			component.auth_type = self.get_auth_type_val (auth_type)
+			key_file = os.path.join (self.key_dir, key_file)
 			if file:
 				in_file = os.path.join(self.inp_dir, file)
 			else:
 				in_file = os.path.join(self.out_dir, component.name.decode() + '.bin')
 				gen_file_with_size (in_file, 0x10)
+				if component.name == mono_sig.encode():
+					component.attribute = COMPONENT_ENTRY._attr['RESERVED']
+					compress_alg        = 'Dummy'
+					is_last_entry       = True
+
+			# compress the component
 			lz_file = compress (in_file, compress_alg, self.out_dir, self.tool_dir)
 			component.data = bytearray(get_file_data (lz_file))
-			key_file = os.path.join (self.key_dir, key_file)
+
+			# calculate the component auth info
 			component.hash_data, component.auth_data = CONTAINER.calculate_auth_data (lz_file, auth_type, key_file, self.out_dir)
 			component.hash_size = len(component.hash_data)
 			if region_size == 0:
+				# arrange the region size automatically
 				region_size = len(component.data)
 				region_size = get_aligned_value (region_size, 4) + len(component.auth_data)
-				region_size = get_aligned_value (region_size, (1 << component.alignment))
+				if  is_mono_signing:
+					region_size = get_aligned_value (region_size, self.header.alignment)
+				else:
+					region_size = get_aligned_value (region_size, (1 << component.alignment))
 			component.size = region_size
+			name_set.add (component.name)
 			self.header.comp_entry.append (component)
 
+		if len(name_set) != len(self.header.comp_entry):
+			raise Exception ("Found duplicated component names in a container !")
+
+		# calculate the component offset based on alignment requirement
 		base_offset = None
 		offset = self.get_header_size ()
 		for component in self.header.comp_entry:
 			alignment = (1 << component.alignment) - 1
-			offset  = (offset + alignment) & ~alignment
+			next_offset  = (offset + alignment) & ~alignment
+			if is_mono_signing and  (next_offset - offset >=  sizeof(LZ_HEADER)):
+				offset = next_offset - sizeof(LZ_HEADER)
+			else:
+				offset = next_offset
 			if base_offset is None:
 				base_offset = offset
 			component.offset = offset - base_offset
 			offset += component.size
 
-		self.adjust_header ()
+		if is_mono_signing:
+			# for monolithic signing, set proper flags and update header
+			self.set_header_flags (CONTAINER_HDR._flags['MONO_SIGNING'])
```

Comment:
A little confused in set_header_flags (overwrite). So, this line does OR operation, right? Not set value.

Reply:
If overwrite is set to 1, it is actually setting the flags again. hence the name.